### PR TITLE
Add Logit Lens and Tuned Lens methods

### LINKED ISSUE #85: Include `LogitLens`

### PR DESCRIPTION
**Pull request by bernasraphael**
## Description

This PR adds Logit Lens and Tuned Lens methods for inspecting predictions decoded from intermediate transformer representations.

Both methods integrate with `ModelWithSplitPoints`: the lens operates on the wrapped model's configured split point and reuses the model's prediction path.
Main changes:

- Add `LogitLens` for projecting an intermediate activation through the model's prediction path.
- Add `TunedLens` with an affine translator tied to the configured split point.
- Add batched fitting, evaluation metrics, and versioned tensor-only Tuned Lens checkpoints.
- Support causal language modeling, masked language modeling, and compatible single-label sequence classification.
- Add conservative automatic projection-path resolution and explicit `head_name`, `pre_head_name`, and `pooling_strategy` overrides.
- Add `plot_lens`, following the existing Interpreto visualization conventions.
- Export the new public API from `interpreto`.
- Add API documentation and an executed notebook covering language modeling, classification, fitting, metrics, and visualization.
- Add focused tests for input validation, batching, padding, projection fidelity, device handling, metrics, checkpoint validation, and visualization alignment.

## Related Issue

#85 

## Type of Change

- 📚 Examples / docs / tutorials / dependencies update
- 🚀 New feature (non-breaking change which adds functionality)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: docs/api/lens/methods/logit_lens.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+# Logit Lens
+
+`LogitLens` follows the method introduced by
+[nostalgebraist](https://www.lesswrong.com/posts/AcKRB8wDpdaN6v6ru/interpreting-gpt-the-logit-lens). It projects the residual
+stream at every model depth through the model's native prediction path. The method has no learned parameters.
+
+The prediction head was trained on final states, so early-layer scores should be interpreted as rankings rather than
+calibrated probabilities. Tuned Lens addresses part of this distribution mismatch by learning a translator for every
+non-final state.
```
Comment (AntoninPoche) [inline docs/api/lens/methods/logit_lens.md:13]: This could be in the class docstring directly.

File: docs/api/lens/methods/tuned_lens.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+---
+
+# Tuned Lens
+
+`TunedLens` follows [Belrose et al. (2023)](https://arxiv.org/abs/2303.08112). It learns one affine residual
+translator for each non-final model state and trains all of them to match the model's final prediction distribution.
+
+The translators start at zero, making an unfitted Tuned Lens exactly equivalent to a Logit Lens. Fitting processes one
+text at a time while projecting all model depths together. Use separate training and evaluation texts when assessing
+the fitted lens.
```
Comment (AntoninPoche) [inline docs/api/lens/methods/tuned_lens.md:12]: I think this should be in the docstring directly. (as the docstring is shown in the doc).

File: docs/api/lens/overview.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+splitter = AllLayersSplitter("hf-internal-testing/tiny-random-gpt2")
+lens = LogitLens(splitter, top_k=3)
+results = lens("Interpreto is useful.")
+```
+
+`results` is ordered like `splitter.activation_names`. Each entry contains `top_indices` and `top_scores`. Language
+model outputs have shape `(1, sequence_length, k)`; sequence-classification outputs have shape `(1, k)`, where `k` is
+`top_k` capped at the model's output size. The singleton dimension represents the one input text. Input batches and
+pre-tokenized inputs are intentionally not part of this initial API.
```
Comment (AntoninPoche) [inline docs/api/lens/overview.md:23]: Does it support a list of strings by computing them iteratively, or is it the user's role to do so?

File: docs/api/lens/overview.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+The scores are normalized over the complete output vocabulary or class set. Intermediate states, particularly early
+ones, do not follow the final prediction head's training distribution, so these values are useful for rankings and
+within-model comparisons rather than as calibrated probabilities.
+
+## Tuned Lens
```
Comment (AntoninPoche) [inline docs/api/lens/overview.md:29]: If you describe both methods here, you could put their api description from mkdocs at the end or in the middle.

Therefore, making it a single file for all the lens modules and simplifying the doc. Your call.

File: docs/api/lens/overview.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+## Visualization
+
+```python
+from interpreto import plot_lens
+
+plot_lens(results, "Interpreto is useful.", tokenizer=splitter.tokenizer)
+```
+
+For sequence classification, pass `label_names` to display readable class names.
```
Comment (AntoninPoche) [inline docs/api/lens/overview.md:56]: You should add the plot line directly using the code snippet above.

File: docs/api/lens/overview.md
Comment (AntoninPoche) [inline docs/api/lens/overview.md:1]: I think that you should put most of the comments directly in the docstrings.

You give two examples, LogitLens and TunedLens. If you put these examples at the beginning, you could do LogitLens + Classification and TunedLens + Generation. Including the plotting lines.

Also, you can add the plotting function mkdocs description at the end.

File: docs/api/visualizations.md [hunk 6]
```diff
@@ -7,3 +7,13 @@
...
+
+## `plot_lens`
+
+`plot_lens` renders the normalized output of `LogitLens.explain()` or `TunedLens.explain()`. Pass the same text used for
+the explanation and the splitter tokenizer. For sequence classification, `label_names` can map label ids to readable
+names. Like the other Interpreto plotting functions, it accepts `custom_css` and an optional `save_path`.
+
+The displayed values are intermediate softmax scores, not calibrated probabilities. Language-model tooltips show the numerical top-k scores at each position. For classification, bar lengths show the corresponding class scores.
+
+::: interpreto.visualizations.plot_lens
```
Comment (AntoninPoche) [inline docs/api/visualizations.md:19]: I think it is more natural at the end of a general lens.md file.

Others should be too, but that is not your problem.

File: docs/api/lens/overview.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+Both methods use [`AllLayersSplitter`](../concepts/splitters/all_layers_splitter.md). It captures the state before the
+first block and after every block in one trace. The lens concatenates those states along the leading dimension and
+passes them through the model's native prediction path together.
+
+```python
+from interpreto import AllLayersSplitter, LogitLens
+
+splitter = AllLayersSplitter("hf-internal-testing/tiny-random-gpt2")
+lens = LogitLens(splitter, top_k=3)
```
Comment (AntoninPoche) [inline docs/api/lens/overview.md:16]: If we force the AllLayersSplitter, should we not hide it in the LogitLens init?

File: interpreto/lens/logit_lens.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        top_k (int): Maximum number of token or class scores returned per prediction.
+
+    Examples:
+        >>> from interpreto import AllLayersSplitter, LogitLens
+        >>> splitter = AllLayersSplitter("hf-internal-testing/tiny-random-gpt2")
+        >>> lens = LogitLens(splitter, top_k=3)
+        >>> results = lens("Interpreto is useful.")
+        >>> list(results) == splitter.activation_names
+        True
+    """
```
Comment (AntoninPoche) [inline interpreto/lens/logit_lens.py:47]: If LogitLens is exactly the Base class, then there is no need for a base class. You can have `TunedLens` inherit directly from `LogitLens`.

File: interpreto/visualizations/commons.py
Comment (AntoninPoche) [inline interpreto/visualizations/commons.py:1]: Did you verify this did not break the other visualizations?

File: interpreto/visualizations/lens.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+_LENS_STYLES = """
+.lens-layer { margin: .8rem 0; border: 1px solid #d8dee9; border-radius: .4rem; }
+.lens-layer summary { cursor: pointer; padding: .6rem; font-weight: 700; background: #f5f7fa; }
+.lens-layer table { width: 100%; border-collapse: collapse; }
+.lens-layer th, .lens-layer td { padding: .45rem .6rem; border-top: 1px solid #e2e8f0; text-align: left; }
+.lens-token { white-space: pre; }
+.lens-score { color: #526172; }
+"""
+
+Tokenizer = PreTrainedTokenizer | PreTrainedTokenizerFast
```
Comment (AntoninPoche) [inline interpreto/visualizations/lens.py:49]: You can use the class `PreTrainedTokenizerBase,` which is the parent of both.

File: mkdocs.yml [hunk 10]
```diff
@@ -61,6 +63,11 @@ nav:
         - Reconstruction Metrics: api/concepts/metrics/reconstruction_metrics.md
         - Sparsity: api/concepts/metrics/sparsity_metrics.md
     - Visualizations: api/visualizations.md
+    - Lens:
+      - Overview: api/lens/overview.md
+      - Methods:
+        - Logit Lens: api/lens/methods/logit_lens.md
+        - Tuned Lens: api/lens/methods/tuned_lens.md
```
Comment (AntoninPoche) [inline mkdocs.yml:70]: I suggested above to merge everything in a single file, no folders. You tell me.

File: README.md
Comment (AntoninPoche) [inline README.md:1]: When you modify the `readme.md`, do not forget the `index.md`

File: interpreto/visualizations/lens.py
Comment (AntoninPoche) [inline interpreto/visualizations/lens.py:1]: I will be better able to comment after the notebook is compiled.

File: docs/notebooks/lens_notebook.ipynb
Comment (AntoninPoche) [inline docs/notebooks/lens_notebook.ipynb:1]: The notebook should be compiled for people to check its results.

You can include your name as the author.

You can add classification.

Do not hesitate to use real models, even if small.

**Review by AntoninPoche — CHANGES_REQUESTED** [review]
Really nice work, much smoother than the previous version!

My comments are mainly on the documentation.

Also, I do not think we need a `BaseLens`, if `LogitLens` correspond to it exactly.
